### PR TITLE
RFC: Embedly link preview indexing

### PR DESCRIPTION
**Pull request by gsidebo**
[Rendered](https://github.com/mitodl/open-discussions/blob/c9a2fc6bf95bafb417801490ac57b36c30726594/docs/rfcs/0002-indexing-embedly-link-preview.md)

#### What are the relevant tickets?
RFC for #1811 

#### What's this PR do?
Adds an RFC for indexing preview content from Embedly to make link posts searchable.

### CODE REVIEW

File: docs/rfcs/0002-indexing-embedly-link-preview.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  "content": "<div><p>NPY acts by sticking to receptor proteins, and the pharmaceutical industry ...",
+  "title": "A New Way to Keep Mosquitoes From Biting",
+  "provider_name": "The Atlantic",
+  ...
+}
+```
+
+The `title` and `provider_name` are potentially significant in terms of the search-ability and relatability of these document (example: if I search "Atlantic", it makes some sense to include this link post from the publication called "The Atlantic" in the results). 
+
+*Proposal*: Prepend `provider_name` and `title` to the text that we index for link posts, and separate the values by dashes (et al). Using the example above, the text that we index would look like this: `"The Atlantic - A New Way to Keep Mosquitoes From Biting - NPY acts by sticking to receptor proteins, and ..."`
```
Comment (Ferdi) [inline docs/rfcs/0002-indexing-embedly-link-preview.md]: @gsidebo I still have the same question:  Do we currently search for the term in the "Title" field of the doc ?

File: docs/rfcs/0002-indexing-embedly-link-preview.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+#### 3) Using the existing `text` model field for storing link preview text
+
+*Proposal*:  Save the preview text for a link post in the `channels.models.Post.text` field (we wouldn't attempt to save text for the corresponding reddit object)
+
+*Potential Issues*: There might still be some application logic on the back- _or_ front-end that considers a post with non-blank `text` value to be a self/text post. This would break that logic.
+
+#### 4) (Possible) Getting rid of either the `plain_text` or `text` field in ES documents
```
Comment (Ferdi) [inline docs/rfcs/0002-indexing-embedly-link-preview.md]: @gsidebo Is this related to saving content from embedly ?

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 1 above)
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: Yes we do currently search for the title field, and it's also incorporated into related post queries

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 2 above)
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: This affects any scenario where we index some text content for a post, which includes indexing this embedly content. It's not specific to embedly. I figured that this probably wasn't a big enough consideration for a completely new RFC and that I could gather feedback on it while we're already thinking about indexing text content. I didn't intend to do this as part of the PR fro #1811. I'll clarify that

File: docs/rfcs/0002-indexing-embedly-link-preview.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+*Potential Issues*: There might still be some application logic on the back- _or_ front-end that considers a post with non-blank `text` value to be a self/text post. This would break that logic.
+
+#### 4) (Related, but separate issue) Getting rid of either the `plain_text` or `text` field in ES documents
+
+After reviewing this a bit, it became clear that we don't seem to be using the ES documents' `text` field. We only care about the plain text content of these posts (stored in the `plain_text` field). 
+
+*Proposal*: (a) Get rid of the `text` field in ES documents, or (b) get rid of `plain_text`, and instead store only plain text content in the `text` field.
+
+*Drawbacks*: For (a) there may still be some application logic that makes use of that value. For (b) it could be jarring to have two properties with the same name on the model object and its corresponding ES document with different values.
```
Comment (rhysyngsun) [inline docs/rfcs/0002-indexing-embedly-link-preview.md]: The frontend definitely expects `text` to be the markdown version and `plain_text` to be what you expect. [See this utility](https://github.com/mitodl/open-discussions/blob/master/static/js/lib/search.js#L57) that converts ES results to the frontend `Post` type. We need to be careful here because there's a few dimensions to this problem:

a) when code is deployed that indexes with the document structure
b) when a full reindex actually occurs
c) when frontend code is deployed that handles a) and b)

So I think we'll need to follow some kind of deprecation process

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 1 above)
Comment (Ferdi) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: Great, than I don't think to do anything for "provider_name" . Title and body/content would be enough

File: docs/rfcs/0002-indexing-embedly-link-preview.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+}
+```
+
+The `title` and `provider_name` are potentially significant in terms of the search-ability and relatability of these document (example: if I search "Atlantic", it makes some sense to include this link post from the publication called "The Atlantic" in the results). 
+
+*Proposal*: Prepend `provider_name` and `title` to the text that we index for link posts, and separate the values by dashes (et al). Using the example above, the text that we index would look like this: `"The Atlantic - A New Way to Keep Mosquitoes From Biting - NPY acts by sticking to receptor proteins, and ..."`
+
+#### 3) Using the existing `text` model field for storing link preview text
+
+*Proposal*:  Save the preview text for a link post in the `channels.models.Post.text` field (we wouldn't attempt to save text for the corresponding reddit object)
```
[the remark this answers is not in the document]
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: 👍 Easy enough

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: The `title` value that I'm referring to is the one that we get from Embedly, which is the actual HTML document title of the page at the URL that you provide. The title that we show in search results and in post summary/detail pages is the title that the user submits. That title value would be unaffected. I'm suggesting that we add the title value to the summary text for the document in ES. We do not currently show that summary text (i.e.: `plain_text`) to the user under any circumstances for link posts, and this issue doesn't include any design changes for link posts in search results

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 3 above)
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: Agreed, will need to be done in stages. I'll add here that the `text` field is never used after converting an ES post result to a frontend `Post` object.

File: docs/rfcs/0002-indexing-embedly-link-preview.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
+## Indexing link preview content from Embedly to make link posts searchable 
+
+Read the description of [#1811](https://github.com/mitodl/open-discussions/issues/1811) for a description of the problem we're solving here. This RFC is to address some decisions about how we go about indexing content from Embedly responses and the index structure in general.
+The goal of this RFC is to call attention to some of the finer points of implementation.
```
[the remark this answers is not in the document]
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: @Ferdi Worth getting your input on this. I'm assuming the answer is "no" for scoping reasons, but if we want this in the future it's good to consider it now. If we want to index the embedly preview for links that a user has included in their article post, I imagine we would either (a) add a new field to the index (`link_content` or something), or (b) intersperse the full preview text in the positions where the user included the links. Either way, we would have documents that could match a search based on text that the user wrote _or_ text from an external site. One case where that would make things interesting is if we ever want to show search results that include highlighted text that matched the search term. We should also consider the fact that an article post can include any number of links.

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsidebo) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: Looks like this won't be needed anyway, so disregard

File: docs/rfcs/0002-indexing-embedly-link-preview.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ferdi) [reply docs/rfcs/0002-indexing-embedly-link-preview.md]: @gsidebo I suggest we keep it simple for now
